scipy/optimize/_zeros_py.py

```diff
@@ -1392,3 +1393,344 @@ def toms748(f, a, b, args=(), k=1,
+            temp_res = _prepare_result(shape, res, active, nit, nfev)
+            if _call_callback_maybe_halt(callback, temp_res):
+                cb_terminate = True
+                break
+        if active.size==0:
+            break
+
+        # [1] Flowchart 4 / Equation 1
+        xi1 = (x1 - x2) / (x3 - x2)
+        phi1 = (f1 - f2) / (f3 - f2)
+        alpha = (x3 - x1) / (x2 - x1)
+        j = ((1 - np.sqrt(1 - xi1)) < phi1) & (phi1 < np.sqrt(xi1))
+
+        f1j, f2j, f3j, alphaj = f1[j], f2[j], f3[j], alpha[j]
+        t = np.full_like(alpha, 0.5)
+        t[j] = (f1j / (f1j - f2j) * f3j / (f3j - f2j)
+                - alphaj * f1j / (f3j - f1j) * f2j / (f2j - f3j))
+
+        # See [1] Appendix, "Adjust T Away from the Interval Boundary"
+        t = np.clip(t, tl, 1-tl)
+
+    res.status[active] = _ECALLBACK if cb_terminate else _ECONVERR
+    return _prepare_result(shape, res, active, nit, nfev)
+
+def _chandrupatla_iv(func, a, b, args, xatol, xrtol,
+                     fatol, frtol, maxiter, callback):
+
+    if not callable(func):
+        raise ValueError('`func` must be callable.')
+
+    # a and b have more complex IV that is taken care of during initialization
+
+    if not np.iterable(args):
+        args = (args,)
+
+    # rudimentary, but enough IMO
+    tols = np.asarray([xatol, xrtol, fatol if fatol is not None else 1, frtol])
+    if (not np.issubdtype(tols.dtype, np.number)
+            or np.any(tols < 0)
+            or tols.shape != (4,)):
+        raise ValueError('Tolerances must be non-negative scalars.')
+
+    maxiter_int = int(maxiter)
+    if maxiter != maxiter_int or maxiter < 0:
+        raise ValueError('`maxiter` must be a non-negative integer.')
+
+    if callback is not None and not callable(callback):
+        raise ValueError('`callback` must be callable.')
+
+    return func, a, b, args, xatol, xrtol, fatol, frtol, maxiter, callback
+
+def _initialize_xf(func, a, b, args):
+    # initializing left and right bracket and function value arrays
+
+    # Try to preserve `dtype`, but we need to ensure that the arguments are at
+    # least floats before passing them into the function because integers
+    # can overflow and cause failure.
+    x1, x2 = np.broadcast_arrays(a, b)  # broadcast and rename
+    xt = np.result_type(x1.dtype, x2.dtype, np.float16)  # at least float16
+    x1, x2 = x1.astype(xt, copy=False)[()], x2.astype(xt, copy=False)[()]
+    f1, f2 = func(x1, *args), func(x2, *args)
+
+    # It's possible that the functions will return multiple outputs for each
+    # scalar input, so we need to broadcast again. All arrays need to be,
+    # writable, so we'll need to copy after broadcasting. Finally, we're going
+    # to be doing operations involving `x1`, `x2`, `f1`, and `f2` throughout,
+    # so figure out the right type from the outset.
+    x1, f1, x2, f2 = np.broadcast_arrays(x1, f1, x2, f2)
+    ft = np.result_type(x1.dtype, x2.dtype, f1.dtype, f2.dtype, np.float16)
```

Comment:
Why restrict the output type with the type of the inputs? You could have a function which does a "reduction" in terms of type.

Actually, what was the reason for looking into the type of input/output. I am not sure I've seen this in much places where we deal with Python only. Not saying it's a bad thing though.

Reply:
We discussed that since there is interpolation, the output type and input type will mix, so we should just make it consistent from the outset.

We also discussed that I am trying to set a precedent for respecting dtypes. This will be important for non-NumPy arrays.